### PR TITLE
[backend] Initial cloud uploader backend script

### PR DESCRIPTION
**Pull request by ChrisBr**
this is the initial prototype which gets called by the backend.

### CODE REVIEW

File: dist/clouduploader.rb [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+end
+
+def upload_image_to_ec2(image, credentials, region, filename)
+  STDOUT.write("Start uploading image #{filename}.\n")
+  name = File.basename(filename)
+
+  Open3.popen2e(
+    'ec2uploadimg',
+    "--description='obs uploader'",
+    '--machine=x86_64',
```
Comment (adrianschroeter) [inline dist/clouduploader.rb]: do we have a failure handling somewhere when it is not a x86_64 image?

**Comment by mdeniz** [comment]
[EC2: Review uploader script](https://trello.com/c/etB3PLXk/1276-ec2-review-uploader-script)

File: dist/clouduploader.rb (hunk 1 above)
Comment (ChrisBr) [reply dist/clouduploader.rb]: We do not know the architecture at this point! I would check the architecture in the frontend and only allow x86_64 instead.

File: dist/clouduploader.rb [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+data_path = ARGV[3]
+filename = ARGV[4]
+
+data = JSON.parse(File.read(data_path))
+DURATION = 60 * 60
+KEY_NAME = 'obs'.freeze
+PRIVATE_KEY = '/etc/clouduploader.pem'.freeze
+HOME = '/etc/obs/cloudupload'.freeze
+ENV['HOME'] = HOME
+ENV['PYTHONUNBUFFERED'] = '1'
```
Comment (bgeuken) [inline dist/clouduploader.rb]: Nitpick: I would prefer to have constants at the top, right after the requires.

**Comment by bgeuken** [comment]
@ChrisBr I would be nice to have a help option for this. Right now one would have to take a look into the script to know the order of the arguments.
It might even make sense to use optionparser here.

Also I would consider providing some validation, eg. validating the region, and add some error handling, eg. when the image upload fails.

**Comment by ChrisBr** [comment]
> I would be nice to have a help option for this. Right now one would have to take a look into the script to know the order of the arguments.

This script is supposed to be only executed by the backend directly and not by a human being. So I'm not sure why we would need a help but I can certainly add one ...

> some error handling, eg. for when the image upload fails.

What should we do when the image upload fails? The only thing we can do here is write it to the log file and provide it to the user (what we already do actually)! That is basically the same as with build jobs, we do not handle build failures, we just provide the log file to the user.

File: dist/clouduploader.rb (hunk 2 above)
Comment (ChrisBr) [reply dist/clouduploader.rb]: I can do that!

File: dist/clouduploader.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  raise 'Wrong number of arguments, please provide: user platform upload_file targetdata filename'
+end
+
+platform = ARGV[1]
+image_path = ARGV[2]
+data_path = ARGV[3]
+filename = ARGV[4]
+
+data = JSON.parse(File.read(data_path))
+DURATION = 60 * 60
```
Comment (hennevogel) [inline dist/clouduploader.rb]: This will always be the same right? :-)

Comment (ChrisBr) [reply dist/clouduploader.rb]: It is the maximum of 1 hour for now. Maybe we can decrease it in the future when we know how long an upload usually takes.

Comment (bgeuken) [reply dist/clouduploader.rb]: Thinking about it... DURATION_IN_SECONDS might be easier to understand.

And Henne means that you could just write 3600 ;-)

**Comment by bgeuken** [comment]
Just checked the card again. It says "robust for wrong input". Is that the case now?

**Comment by hennevogel** [comment]
> robust for wrong input

What was meant by this is: `doesn't let you become root if you give it a well crafted image binary.`

**Comment by ChrisBr** [comment]
> What was meant by this is: doesn't let you become root if you give it a well crafted image binary.

Yep, also when you provide e.g. an image name or arn with ``arn:;rm -rf``. This can not happen as we use open3 library.

Wrong input in terms of region etc. we should handle in the frontend. The script will just fail and write it to the log which the user needs to examine.

**Comment by bgeuken** [comment]
Good, that was all I wanted to make sure.

I didn't ask for the region validation, etc. in this context;-)

**Pull request merged**